File: src/lsp/schema.go
// The schema reader and the note checker. A file in spec/schemas says what one
// kind of note holds, this reads it, and the checker weighs a note against it.
// The caller hands the tree in, so a test drives both over a fake one.
// [[spec/design_output/schema#the-reader-and-the-checker]]
package main

import (
	"quackitect/yaml"

	"fmt"
	"sort"
	"strings"
)

const (
	Schemas   = "spec/schemas"
	SchemaEnd = ".schema.yaml"
	shown     = 40
	ellipsis  = "..."
)

// [[spec/design_output/schema#the-schemas-read-once]]
type Kinds struct {
	order []string
	at    map[string]*yaml.Doc
}

func (one *Kinds) set(kind string, said *yaml.Doc) {
	if one.at == nil {
		one.at = map[string]*yaml.Doc{}
	}
	if _, held := one.at[kind]; !held {
		one.order = append(one.order, kind)
	}
	one.at[kind] = said
}

func (one *Kinds) Get(kind string) *yaml.Doc {
	if one == nil {
		return nil
	}
	return one.at[kind]
}

func (one *Kinds) Len() int {
	if one == nil {
		return 0
	}
	return len(one.order)
}

func (one *Kinds) Names() []string {
	out := append([]string(nil), one.order...)
	sort.Strings(out)
	return out
}

// [[spec/design_output/schema#a-schema-names-its-chapters]]
func isNoteSchema(said *yaml.Doc) bool {
	if said == nil || yaml.AsString(said.Get("kind")) == "" {
		return false
	}
	body := yaml.AsDoc(said.Get("body"))
	return body != nil && len(yaml.AsList(body.Get("sections"))) > 0
}

// [[spec/design_output/schema#the-schemas-read-once]]
func schemasIn(tree *Tree) *Kinds {
	out := &Kinds{}
	for _, name := range tree.Names(Schemas, SchemaEnd) {
		said := yaml.AsDoc(yaml.Read(tree.Read(Schemas + "/" + name)))
		if isNoteSchema(said) {
			out.set(yaml.AsString(said.Get("kind")), said)
		}
	}
	return out
}

// [[spec/design_output/schema#a-folder-names-its-kind]]
func governorOf(schemas *Kinds, path string) *yaml.Doc {
	where := slashed(path)
	for _, kind := range schemas.order {
		schema := schemas.at[kind]
		for _, glob := range yaml.StringsOf(schema.Get("governs")) {
			if matches(glob, where) {
				return schema
			}
		}
	}
	return nil
}

// [[spec/design_output/schema#a-folder-names-its-kind]]
func strangerFault(text string, schema *yaml.Doc, where string) (Finding, bool) {
	held := yaml.AsString(schema.Get("kind"))
	kind := kindOf(text)
	if held == "" || kind == held {
		return Finding{}, false
	}
	if kind == "" {
		return schemaFault("Kind", where, 1,
			fmt.Sprintf("%s names no kind, and the %s schema governs this path.", where, held)), true
	}
	return schemaFault("Kind", where, 1,
		fmt.Sprintf("%s reads as a %s, and the %s schema governs this path.", where, kind, held)), true
}

// [[spec/design_output/schema#a-finding-names-the-section]]
// A reading past one buffer wants the tree, and the write door holds none. [[spec/design_output/lsp#a-marked-rule-wants-argument]]
func checkNoteIn(tree *Tree, text string, schema *yaml.Doc, where string) []Finding {
	out := checkNote(text, schema, where)
	return append(out, markedFaults(tree, readNote(text), schema, where)...)
}

func checkNote(text string, schema *yaml.Doc, where string) []Finding {
	note := readNote(text)
	kind := yaml.AsString(schema.Get("kind"))
	out := frontFaults(note, yaml.AsDoc(schema.Get("frontmatter")), kind, where)
	return append(out, bodyFaults(note, yaml.AsDoc(schema.Get("body")), kind, where)...)
}

func frontFaults(note Note, spec *yaml.Doc, kind, where string) []Finding {
	if !note.Front.Stands {
		return []Finding{schemaFault("Frontmatter", where, 1,
			fmt.Sprintf("A %s note opens with frontmatter.", kind))}
	}

	out := []Finding{}
	said := note.Front.Said
	props := yaml.AsDoc(spec.Get("properties"))

	for _, key := range yaml.StringsOf(spec.Get("required")) {
		if !yaml.Empty(said.Get(key)) || waitsForFill(said, yaml.AsDoc(props.Get(key))) {
			continue
		}
		out = append(out, schemaFault(key, where, 1,
			fmt.Sprintf("A %s note names %s in its frontmatter.", kind, key)))
	}

	for _, key := range said.Keys() {
		value := said.Get(key)
		rule := yaml.AsDoc(props.Get(key))
		line := note.Front.Lines[key]
		if line == 0 {
			line = 1
		}
		if rule == nil {
			if spec.Get("additionalProperties") == false {
				out = append(out, schemaFault(key, where, line,
					fmt.Sprintf("The %s schema names no %s.", kind, key)))
			}
			continue
		}
		out = append(out, fieldFaults(key, value, rule, kind, where, line)...)
	}
	return out
}

func fieldFaults(key string, value any, rule *yaml.Doc, kind, where string, line int) []Finding {
	out := []Finding{}
	isLink := yaml.AsBool(rule.Get("x-link"))
	said := value
	if isLink {
		said = linkless(value)
	}

	if isLink && !linked(value) {
		out = append(out, schemaFault(key, where, line,
			fmt.Sprintf("%s names a link, and this reads %s.", key, show(value))))
	}
	if rule.Has("const") && !same(said, rule.Get("const")) {
		out = append(out, schemaFault(key, where, line,
			fmt.Sprintf("%s reads %s, and a %s note names %s.", key, show(said), kind, yaml.AsString(rule.Get("const")))))
	}
	if allowed, held := rule.Get("enum").([]any); held {
		if !holds(allowed, said) {
			out = append(out, schemaFault(key, where, line,
				fmt.Sprintf("%s reads %s, and the schema allows %s.", key, show(said), joined(allowed, ", "))))
		}
	}
	if rule.Has("type") && !typed(value, rule.Get("type")) {
		out = append(out, schemaFault(key, where, line,
			fmt.Sprintf("%s takes %s, and this reads %s.", key, joined(yaml.Flat(rule.Get("type")), " or "), typeOf(value))))
	}
	return out
}

// [[spec/design_output/schema#a-placeholder-stands-at-warning]]
func placeholderFaults(text string, schema *yaml.Doc, where string) []Finding {
	rows := yaml.SplitLines(text)
	note := readNote(text)
	props := yaml.AsDoc(yaml.AsDoc(schema.Get("frontmatter")).Get("properties"))
	out := []Finding{}

	for _, key := range note.Front.LineKeys {
		line := note.Front.Lines[key]
		rule := yaml.AsDoc(props.Get(key))
		if rule == nil || rule.Has("const") {
			continue
		}
		if _, listed := rule.Get("enum").([]any); listed {
			continue
		}
		if line-1 >= len(rows) || strings.TrimSpace(rows[line-1]) != fmt.Sprintf("%s: %s", key, minted(rule)) {
			continue
		}
		out = append(out, left(where, line, key))
	}

	named := map[string]string{}
	for _, one := range yaml.AsList(yaml.AsDoc(schema.Get("body")).Get("sections")) {
		rule := yaml.AsDoc(one)
		if rule == nil {
			continue
		}
		// A chapter the schema marks x-fills false stays at its comment with no warning. [[spec/design_output/schema#a-placeholder-stands-at-warning]]
		if fills, set := rule.Get("x-fills").(bool); set && !fills {
			continue
		}
		if said := yaml.AsString(rule.Get("description")); said != "" {
			named[yaml.AsString(rule.Get("header"))] = fmt.Sprintf("<!-- %s -->", said)
		}
	}
	for _, held := range note.Sections {
		said, ours := named[held.Header]
		if !ours {
			continue
		}
		for i, line := range held.Own {
			if strings.TrimSpace(line) != said {
				continue
			}
			out = append(out, left(where, held.Line+i+1, held.Header))
			break
		}
	}
	return out
}

// A governed folder holds its own kind alone, in the words folderFault in .claude/skills/level0/lib/schema.js answers. [[spec/tickets/each-folder-holds-its-kind]]
func folderFault(where string, schema *yaml.Doc) Finding {
	kind := yaml.AsString(schema.Get("kind"))
	found := schemaFault("Folder", where, 1,
		fmt.Sprintf("%s stands in a folder the %s schema governs, which holds %s notes alone. Move it off the governed folder.", where, kind, kind))
	found.Severity = SeverityWarning
	return found
}

func left(file string, line int, what string) Finding {
	return Finding{
		File:     file,
		Rule:     "Schema.Placeholder",
		Line:     line,
		Column:   1,
		Message:  what + " still carries the placeholder mint writes. Say what stands there.",
		Severity: SeverityWarning,
	}
}

// [[spec/design_output/schema#the-sweep-over-the-tree]]
func schemaFaults(tree *Tree) []Finding {
	schemas := schemasIn(tree)
	out := []Finding{}
	if schemas.Len() == 0 {
		return out
	}

	for _, path := range tree.Paths() {
		if !strings.HasSuffix(path, ".md") {
			if governor := governorOf(schemas, path); governor != nil {
				out = append(out, folderFault(path, governor))
			}
			continue
		}
		text := tree.Read(path)
		out = append(out, noteFaults(tree, schemas, path, text)...)
	}
	return out
}

// [[spec/design_output/schema#the-sweep-over-the-tree]]
func noteFaults(tree *Tree, schemas *Kinds, path, text string) []Finding {
	out := []Finding{}
	kind := kindOf(text)
	governor := governorOf(schemas, path)

	if governor != nil && yaml.AsString(governor.Get("kind")) != kind {
		if found, stands := strangerFault(text, governor, path); stands {
			out = append(out, found)
		}
		return out
	}
	if kind == "" {
		return out
	}

	schema := schemas.Get(kind)
	if schema == nil {
		return append(out, schemaFault("Kind", path, 1,
			fmt.Sprintf("%s names no schema, and %s holds %s.", kind, Schemas, strings.Join(schemas.Names(), ", "))))
	}
	out = append(out, checkNoteIn(tree, text, schema, path)...)
	return append(out, placeholderFaults(text, schema, path)...)
}

func minted(rule *yaml.Doc) string {
	if rule.Has("const") {
		if yaml.AsBool(rule.Get("x-link")) {
			return fmt.Sprintf("[[%s]]", yaml.AsString(rule.Get("const")))
		}
		return yaml.AsString(rule.Get("const"))
	}
	if allowed, listed := rule.Get("enum").([]any); listed && len(allowed) > 0 {
		return yaml.AsString(allowed[0])
	}

	said := yaml.AsString(rule.Get("description"))
	if said == "" {
		said = "what goes here"
	}
	for _, one := range yaml.Flat(rule.Get("type")) {
		if yaml.AsString(one) == "array" {
			return fmt.Sprintf("[%q]", said)
		}
	}
	if yaml.AsBool(rule.Get("x-link")) {
		return fmt.Sprintf("[[%s]]", said)
	}
	return said
}

func schemaFault(rule, file string, line int, message string) Finding {
	return fault("Schema."+nameOf(rule), file, line, message)
}

func nameOf(said string) string {
	parts := []string{}
	for _, one := range strings.FieldsFunc(said, func(r rune) bool {
		return !(r >= 'a' && r <= 'z' || r >= 'A' && r <= 'Z' || r >= '0' && r <= '9')
	}) {
		parts = append(parts, one)
	}
	if len(parts) == 1 {
		return parts[0]
	}
	out := &strings.Builder{}
	for _, one := range parts {
		out.WriteString(strings.ToUpper(one[:1]) + one[1:])
	}
	return out.String()
}

func headed(standing []standingAt, header string) int {
	for i, one := range standing {
		if one.Header == header {
			return i
		}
	}
	return -1
}

func at(order []string, said string) int {
	for i, one := range order {
		if one == said {
			return i
		}
	}
	return -1
}

func someUnnumbered(items []item) bool {
	for _, one := range items {
		if !numberedAt.MatchString(one.said) {
			return true
		}
	}
	return false
}

func holds(allowed []any, said any) bool {
	for _, one := range allowed {
		if same(one, said) {
			return true
		}
	}
	return false
}

// [[spec/design_output/schema#a-finding-names-the-section]]
func same(a, b any) bool {
	one, ours := a.([]any)
	two, theirs := b.([]any)
	if ours || theirs {
		if !ours || !theirs || len(one) != len(two) {
			return false
		}
		for i := range one {
			if !same(one[i], two[i]) {
				return false
			}
		}
		return true
	}
	if yaml.AsDoc(a) != nil || yaml.AsDoc(b) != nil {
		return false
	}
	return fmt.Sprintf("%T:%v", a, a) == fmt.Sprintf("%T:%v", b, b)
}

func typed(value any, said any) bool {
	for _, one := range yaml.Flat(said) {
		switch yaml.AsString(one) {
		case "array":
			if _, held := value.([]any); held {
				return true
			}
		case "object":
			if yaml.AsDoc(value) != nil {
				return true
			}
		case "string":
			if _, held := value.(string); held {
				return true
			}
		case "integer", "number":
			if _, held := value.(int); held {
				return true
			}
		case "boolean":
			if _, held := value.(bool); held {
				return true
			}
		default:
			return true
		}
	}
	return false
}

func typeOf(value any) string {
	if _, held := value.([]any); held {
		return "a list"
	}
	if yaml.AsDoc(value) != nil {
		return "a map"
	}
	return "one line"
}

func show(said any) string {
	flatSaid := yaml.AsString(said)
	if one, held := said.([]any); held {
		flatSaid = joined(one, ", ")
	}
	if len(flatSaid) > shown {
		return flatSaid[:shown-len(ellipsis)] + ellipsis
	}
	return flatSaid
}

func joined(said []any, with string) string {
	parts := make([]string, 0, len(said))
	for _, one := range said {
		parts = append(parts, yaml.AsString(one))
	}
	return strings.Join(parts, with)
}

// [[spec/design_output/schema#a-finding-names-the-section]]
func waitsForFill(said, rule *yaml.Doc) bool {
	if rule == nil {
		return false
	}
	from := yaml.AsString(rule.Get("x-filled-by"))
	return from != "" && said.Has(from) && yaml.Empty(said.Get(from))
}

File: .claude/skills/level0/lib/schema.js
// The schema reader and the note checker. A file in spec/schemas says what one
// kind of note holds, this reads it, and the checker weighs a note against it.
// The caller hands the tree in, so a test drives both over a fake one.
// [[spec/design_output/schema#the-reader-and-the-checker]]

import { matches } from "./paths.js";
import { bodyFaults, placeholderFaults } from "./schema-body.js";
import {
  empty,
  fault,
  LEFT,
  lineOf,
  linked,
  linkless,
  show,
  solved,
  typed,
  typeOf,
} from "./schema-fault.js";
import { kindOf, readNote } from "./schema-read.js";
import { refersFaults, slotFaults } from "./schema-route.js";
import { keyed, readYaml } from "./schema-yaml.js";

export {
  bodyFaults,
  CHECKED,
  chaptersWanted,
  itemsIn,
  left,
  minted,
  placeholderFaults,
} from "./schema-body.js";
export {
  empty,
  fault,
  LEFT,
  LINK,
  lineOf,
  linked,
  linkless,
  refOf,
  SEVERITY,
  show,
  solved,
  typed,
  typeOf,
} from "./schema-fault.js";
export { FENCE, kindOf, readNote, sectionAt } from "./schema-read.js";
export {
  canonicalOf,
  entriesIn,
  entryNamed,
  hashOf,
  names,
  OUTSIDE,
  processHash,
  refersFaults,
  slotFaults,
  wanted,
} from "./schema-route.js";

export { readYaml };

export const SCHEMAS = "spec/schemas";

export const END = ".schema.yaml";

export const MINT_TOOL = "mint_note";

// [[spec/design_output/schema#a-schema-names-its-chapters]]
export function isNoteSchema(said) {
  return Boolean(said?.kind) && Boolean(said?.body?.sections?.length);
}

// [[spec/design_output/schema#a-data-schema-holds-yaml]]
export function isDataSchema(said) {
  return Boolean(said?.kind) && Boolean(said?.data);
}

// [[spec/design_output/schema#the-schemas-read-once]]
export function schemasIn(tree) {
  return kindsIn(tree, isNoteSchema);
}

// [[spec/design_output/schema#a-data-schema-holds-yaml]]
export function dataSchemasIn(tree) {
  return kindsIn(tree, isDataSchema);
}

// [[spec/design_output/schema#one-home-for-a-shape]]
export function allSchemasIn(tree) {
  return kindsIn(tree, (said) => Boolean(said?.kind));
}

function kindsIn(tree, holds) {
  const out = new Map();
  for (const name of tree.names(SCHEMAS, END)) {
    const said = readYaml(tree.read(`${SCHEMAS}/${name}`));
    if (holds(said)) out.set(String(said.kind), said);
  }
  return out;
}

// [[spec/design_output/schema#a-folder-names-its-kind]]
export function governorOf(schemas, path) {
  const where = String(path ?? "")
    .split("\\")
    .join("/");
  for (const schema of schemas?.values?.() ?? []) {
    for (const glob of [schema.governs ?? []].flat()) {
      if (matches(glob, where)) return schema;
    }
  }
  return null;
}

// [[spec/design_output/schema#a-folder-names-its-kind]]
export function strangerFault(text, schema, where) {
  const held = String(schema?.kind ?? "");
  const kind = kindOf(text);
  if (!held || kind === held) return null;
  if (!kind) {
    return fault(
      "Kind",
      where,
      1,
      `${where} names no kind, and the ${held} schema governs this path.`,
    );
  }
  return fault(
    "Kind",
    where,
    1,
    `${where} reads as a ${kind}, and the ${held} schema governs this path.`,
  );
}

// [[spec/design_output/schema#a-finding-names-the-section]]
export function checkNote(text, schema, where, schemas) {
  const note = readNote(text);
  const kind = String(schema?.kind ?? "");
  const spec = schema?.frontmatter ?? {};
  if (!note.front.stands) {
    return [fault("Frontmatter", where, 1, `A ${kind} note opens with frontmatter.`)];
  }
  const held = heldIn(note.front, schema, kind, where, "frontmatter", schemas);
  return [
    ...mapFaults(note.front.said ?? {}, spec, held, ""),
    ...slotFaults(note.front.said ?? {}, where, note.front.lines),
    ...bodyFaults(note, schema?.body ?? {}, kind, where),
  ];
}

// [[spec/design_output/schema#a-data-schema-holds-yaml]]
export function checkData(text, schema, where, schemas) {
  const lines = new Map();
  const said = readYaml(text, lines);
  const spec = schema?.data ?? {};
  const kind = String(schema?.kind ?? "");
  const front = { said, lines: Object.fromEntries(lines) };
  return [
    ...mapFaults(said, spec, heldIn(front, schema, kind, where, "file", schemas), ""),
    ...slotFaults(said, where, front.lines),
  ];
}

// [[spec/design_output/schema#the-checker-walks-every-key]]
function heldIn(front, schema, kind, where, calls, schemas) {
  return {
    kind,
    where,
    lines: front.lines ?? {},
    root: front.said ?? {},
    calls,
    schema,
    schemas,
  };
}

// [[spec/design_output/schema#a-finding-names-the-section]]
function waitsForFill(said, rule) {
  const from = rule?.["x-filled-by"];
  return Boolean(from) && Object.hasOwn(said ?? {}, from) && empty(said[from]);
}

// [[spec/design_output/schema#the-checker-walks-every-key]]
function mapFaults(said, spec, held, path) {
  const out = [];
  const props = spec.properties ?? {};

  for (const key of spec.required ?? []) {
    if (!empty(said?.[key]) || waitsForFill(said, props[key])) continue;
    out.push(
      fault(
        key,
        held.where,
        lineOf(held, path),
        path
          ? `${path} names ${key}, and a ${held.kind} names it on every entry.`
          : `A ${held.kind} names ${key} in its ${held.calls}.`,
      ),
    );
  }

  for (const [key, value] of Object.entries(said ?? {})) {
    const at = keyed(path, key);
    const rule = props[key];
    const line = lineOf(held, at);
    if (!rule) {
      if (spec.additionalProperties === false) {
        out.push(
          fault(
            key,
            held.where,
            line,
            path
              ? `The ${held.kind} schema names no ${key} under ${path}.`
              : `The ${held.kind} schema names no ${key}.`,
          ),
        );
      }
      continue;
    }
    out.push(...fieldFaults(key, value, solved(rule, held), held, at, line));
  }
  return out;
}

function fieldFaults(key, value, rule, held, at, line) {
  const out = [];
  const kind = held.kind;
  const where = held.where;
  const said = rule["x-link"] ? linkless(value) : value;

  if (rule["x-link"] && !linked(value)) {
    out.push(
      fault(key, where, line, `${key} names a link, and this reads ${show(value)}.`),
    );
  }
  if (rule.const !== undefined && said !== rule.const) {
    out.push(
      fault(
        key,
        where,
        line,
        `${key} reads ${show(said)}, and a ${kind} note names ${rule.const}.`,
      ),
    );
  }
  if (Array.isArray(rule.enum) && !rule.enum.includes(said)) {
    out.push(
      fault(
        key,
        where,
        line,
        `${key} reads ${show(said)}, and the schema allows ${rule.enum.join(", ")}.`,
      ),
    );
  }
  if (rule.type && !typed(value, rule.type)) {
    out.push(
      fault(
        key,
        where,
        line,
        `${key} takes ${[rule.type].flat().join(" or ")}, and this reads ${typeOf(value)}.`,
      ),
    );
  }
  out.push(...refersFaults(key, value, rule, held, at, line));
  out.push(...deeperFaults(value, rule, held, at));
  return out;
}

// [[spec/design_output/schema#the-checker-walks-every-key]]
function deeperFaults(value, rule, held, at) {
  const out = [];
  const items = solved(rule.items, held);
  if (items?.properties && Array.isArray(value)) {
    for (let i = 0; i < value.length; i++) {
      if (!value[i] || typeof value[i] !== "object") continue;
      out.push(...mapFaults(value[i], items, held, `${at}[${i}]`));
    }
  }
  if (rule.properties && value && typeof value === "object" && !Array.isArray(value)) {
    out.push(...mapFaults(value, rule, held, at));
  }
  return out;
}

// [[spec/design_output/schema#the-sweep-over-the-tree]]
export function schemaFaults(tree) {
  const schemas = schemasIn(tree);
  const data = dataSchemasIn(tree);
  const every = allSchemasIn(tree);
  const out = [];
  if (!schemas.size) return out;

  for (const path of tree.paths()) {
    if (!path.endsWith(".md")) {
      const governor = governorOf(data, path);
      if (governor) out.push(...checkData(tree.read(path), governor, path, every));
      else if (governorOf(schemas, path)) out.push(folderFault(path, governorOf(schemas, path)));
      continue;
    }
    const text = tree.read(path);
    const kind = kindOf(text);
    const governor = governorOf(schemas, path);

    if (governor && governor.kind !== kind) {
      out.push(strangerFault(text, governor, path));
      continue;
    }
    if (!kind) continue;

    const schema = schemas.get(kind);
    if (!schema) {
      out.push(
        fault(
          "Kind",
          path,
          1,
          `${kind} names no schema, and ${SCHEMAS} holds ${[...schemas.keys()].sort().join(", ")}.`,
        ),
      );
      continue;
    }
    out.push(...checkNote(text, schema, path, every));
    out.push(...placeholderFaults(text, schema, path));
  }
  return out;
}

// A governed folder holds its own kind alone, so a page or a picture there stands at warning until the owner moves it. [[spec/tickets/each-folder-holds-its-kind]]
export function folderFault(where, schema) {
  const kind = String(schema?.kind ?? "");
  return {
    ...fault(
      "Folder",
      where,
      1,
      `${where} stands in a folder the ${kind} schema governs, which holds ${kind} notes alone. Move it off the governed folder.`,
    ),
    severity: LEFT,
  };
}

// [[spec/design_output/schema#the-door-refuses-a-departure]]
export function schemasFrom(files) {
  return kindsFrom(files, isNoteSchema);
}

// [[spec/design_output/schema#a-data-schema-holds-yaml]]
export function dataSchemasFrom(files) {
  return kindsFrom(files, isDataSchema);
}

// [[spec/design_output/schema#one-home-for-a-shape]]
export function allSchemasFrom(files) {
  return kindsFrom(files, (said) => Boolean(said?.kind));
}

function kindsFrom(files, holds) {
  const out = new Map();
  for (const one of files ?? []) {
    const said = readYaml(one.text);
    if (holds(said)) out.set(String(said.kind), said);
  }
  return out;
}

export function refusedNote(where, kind, found) {
  return [
    `The ${kind} schema refuses this write to ${where}.`,
    "",
    ...found.map(
      (one) => `  ${where}:${one.line}:${one.column}  ${one.rule}\n    ${one.message}`,
    ),
    "",
    `Run ./RUNME.sh mint ${kind} <path> for the shape it names, or park a draft as _name.md.`,
  ].join("\n");
}

// [[spec/design_output/schema#a-folder-names-its-kind]]
export function refusedKind(where, schema, found) {
  const kind = String(schema?.kind ?? "");
  return [
    `${SCHEMAS}/${kind}${END} governs ${where}, and it refuses this write.`,
    "",
    `  ${where}:${found.line}:${found.column}  ${found.rule}`,
    `    ${found.message}`,
    "",
    `Call ${MINT_TOOL} with kind ${kind}, this path and the fields, and it writes the note.`,
    "A draft named _name.md stands outside every rule while a kind settles.",
  ].join("\n");
}
